File: hena/tyama_hena24_enum.pas
// http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
// http://nabetani.sakura.ne.jp/hena/ord24eliseq/

program hena24;
uses
	System;
var
	i: longint;
	g: sequence of longint;
	s: string;
	z: IEnumerator<longint>;

function isqrt(n: longint): longint;
var
	x: longint;
	y: longint;
begin
	if(n<=0) then
		isqrt:=0
	else if(n<4) then
		isqrt:=1
	else begin
		x:=0;
		y:=n;
		while((x<>y) and (x+1<>y)) do begin
			x:=y;
			y:=(n div y+y) div 2;
		end;
		isqrt:=x;
	end;
end;

function icbrt(n: longint): longint;
var
	x: longint;
	y: longint;
begin
	if(n<0) then
		icbrt:=0
	else if(n<8) then
		icbrt:=1
	else begin
		x:=0;
		y:=n;
		while((x<>y) and (x+1<>y)) do begin
			x:=y;
			y:=(n div y div y+y*2) div 3;
		end;
		icbrt:=x;
	end;
end;

function is_sq(n: longint): boolean;
var
	x: longint;
begin
	x:=isqrt(n);
	is_sq:=x*x=n;
end;

function is_cb(n: longint): boolean;
var
	x: longint;
begin
	x:=icbrt(n);
	is_cb:=x*x*x=n;
end;

function is_multiple(i:longint; n: longint): boolean;
begin
	is_multiple:=i mod n=0;
end;

function is_le(i:longint; n: longint): boolean;
begin
	is_le:=i<=n;
end;

function generate(): sequence of longint;
var
	i: longint;
begin
	i:=1;
	while true do begin
		yield i;
		i:=i+1;
	end;
end;

function drop_prev(check: function(x:longint):boolean; _prev: sequence of longint): sequence of longint;
var
	a: longint;
	b: longint;
	prev: IEnumerator<longint>;
begin
	prev:=_prev.GetEnumerator();
	prev.MoveNext();
	a:=prev.Current;
	prev.MoveNext();
	b:=prev.Current;
	while true do begin
		if not check(b) then yield a;
		a:=b;
		prev.MoveNext();
		b:=prev.Current;
	end;
end;

function drop_next(check: function(x:longint):boolean; _prev: sequence of longint): sequence of longint;
var
	a: longint;
	b: longint;
	prev: IEnumerator<longint>;
begin
	prev:=_prev.GetEnumerator();
	prev.MoveNext();
	a:=prev.Current;
	prev.MoveNext();
	b:=prev.Current;
	yield a;
	while true do begin
		if not check(a) then yield b;
		a:=b;
		prev.MoveNext();
		b:=prev.Current;
	end;
end;

function drop_n(check: function(x:longint;y:longint):boolean; n: longint; _prev: sequence of longint): sequence of longint;
var
	a: longint;
	i: longint;
	prev: IEnumerator<longint>;
begin
	prev:=_prev.GetEnumerator();
	i:=0;
	while true do begin
		i:=i+1;
		prev.MoveNext();
		a:=prev.Current;
		if not check(i,n) then yield a;
	end;
end;

begin
	var f:=new Dictionary<char,Func<sequence of longint,sequence of longint>>();
	f['S']:=x->drop_next(is_sq,x);
	f['s']:=x->drop_prev(is_sq,x);
	f['C']:=x->drop_next(is_cb,x);
	f['c']:=x->drop_prev(is_cb,x);
	f['h']:=x->drop_n(is_le,100,x);
	for i:=2 to 9 do begin
		var j:=i;
		f[char(48+j)]:=x->drop_n(is_multiple,j,x);
	end;

	while true do begin
		s:=Console.ReadLine();
		if s=nil then break;
		g:=generate();
		for i:=1 to length(s) do g:=f[s[i]](g);
		z:=g.GetEnumerator();
		for i:=0 to 9 do begin
			if i>0 then write(',');
			z.MoveNext();
			write(z.Current);
		end;
		writeln();
		Console.Out.Flush();
	end;
end.

File: hena/tyama_hena24_enum.swift
//usr/bin/env swift $0 $@;exit

// http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
// http://nabetani.sakura.ne.jp/hena/ord24eliseq/

precedencegroup PowerPrecedence {
	associativity: right
	higherThan: MultiplicationPrecedence
}
infix operator <>: ComparisonPrecedence
infix operator ≠: ComparisonPrecedence
infix operator ≤: ComparisonPrecedence
infix operator ≥: ComparisonPrecedence
infix operator **: PowerPrecedence
public func <> <T: Equatable>(lhs: T, rhs: T) -> Bool{return lhs != rhs}
public func ≠ <T: Equatable>(lhs: T, rhs: T) -> Bool{return lhs != rhs}
public func ≤ <T: Comparable>(lhs: T, rhs: T) -> Bool{return lhs <= rhs}
public func ≥ <T: Comparable>(lhs: T, rhs: T) -> Bool{return lhs >= rhs}
public func ** <T: Numeric>(lhs: T, rhs: Int64) -> T{
	var x = lhs
	var k = rhs
	var e = 1 as T
	while k>0{
		if k%2>0 {e*=x}
		x*=x
		k/=2
	}
	return e
}

func isqrt(_ n: Int64) -> Int64{
	if n<=0 {return 0}
	if n<4 {return 1}
	var x: Int64 = 0
	var y: Int64 = n
	while x<>y && x+1<>y {
		x = y
		y = (n/y+y)/2
	}
	return x
}
func icbrt(_ n: Int64) -> Int64{
	if n<0 {return icbrt(-n)}
	if n==0 {return 0}
	if n<8 {return 1}
	var x: Int64 = 0
	var y: Int64 = n
	while x<>y && x+1<>y {
		x = y
		y = (n/y/y+y*2)/3
	}
	return x
}

class generate: Sequence, IteratorProtocol{
	var i: Int64 = 0
	func next() -> Int64?{
		i += 1
		return i
	}
}
class drop_prev: Sequence, IteratorProtocol{
	var prev: AnyIterator<Int64>
	var check: (Int64) -> Bool
	var a: Int64 = 0
	var b: Int64 = 0
	init(_ check_: @escaping (Int64) -> Bool, _ prev_: AnyIterator<Int64>){
		check = check_
		prev = prev_
		b = prev.next()!
	}
	func next() -> Int64?{
		while true{
			a = b
			b = prev.next()!
			if !check(b){
				return a
			}
		}
	}
}
class drop_next: Sequence, IteratorProtocol{
	var prev: AnyIterator<Int64>
	var check: (Int64) -> Bool
	var first: Bool = true
	var a: Int64 = 0
	var b: Int64 = 0
	init(_ check_: @escaping (Int64) -> Bool, _ prev_: AnyIterator<Int64>){
		check = check_
		prev = prev_
	}
	func next() -> Int64?{
		while true{
			a = b
			b = prev.next()!
			if first || !check(a){
				first = false
				return b
			}
		}
	}
}
class drop_n: Sequence, IteratorProtocol{
	var prev: AnyIterator<Int64>
	var check: (Int64,Int64) -> Bool
	var n: Int64
	var i: Int64 = 0
	init(_ check_: @escaping (Int64,Int64) -> Bool, _ n_: Int64, _ prev_: AnyIterator<Int64>){
		check = check_
		prev = prev_
		n = n_
	}
	func next() -> Int64?{
		while true{
			i += 1
			let a = prev.next()!
			if !check(i,n){
				return a
			}
		}
	}
}
	
func is_sq(_ n: Int64) -> Bool{return isqrt(n)**2==n}
func is_cb(_ n: Int64) -> Bool{return icbrt(n)**3==n}
func is_multiple(_ i: Int64, _ n: Int64) -> Bool{return i%n==0}
func is_le(_ i: Int64, _ n: Int64) -> Bool{return i<=n}
var f: [Character: (AnyIterator<Int64>) -> AnyIterator<Int64>] = [:]
f["S"]={AnyIterator(drop_next(is_sq,$0))}
f["s"]={AnyIterator(drop_prev(is_sq,$0))}
f["C"]={AnyIterator(drop_next(is_cb,$0))}
f["c"]={AnyIterator(drop_prev(is_cb,$0))}
f["h"]={AnyIterator(drop_n(is_le,100,$0))}
for i in 2...9{f[Character(String(i))]={AnyIterator(drop_n(is_multiple,Int64(i),$0))}}

while true{
	let s = readLine()
	if s==nil{break}
	var g: AnyIterator<Int64> = AnyIterator(generate())
	for c in s!{
		g = f[c]!(g)
	}
	var c = 0
	for e in g{
		c += 1
		print(e,terminator:c==10 ? "\n" : ",")
		if c==10{break}
	}
}